TypeScript Rewrite

elijahr2411 opened the pull request:
Major rewrite-in-progress of the entire backend of the client, now in TypeScript and with the CollabVM portions abstracted from the DOM.

resolves #14 
fixes #15

elijahr2411 commented:
fixes #15

MDMCK10 reviewed (changes requested):
Some things are still a bit broken:

1. build times are horrendous, and it uses stupid amounts of memory/cpu/disk while building
![image](https://github.com/computernewb/collab-vm-1.2-webapp/assets/21245760/c2a1a7d1-764c-4f43-a8f2-c9ff768b35f3)

2. username does not persist (test: change username, then refresh)

3. staff actions don't show up after logging in, and logging in makes username invisible
![image](https://github.com/computernewb/collab-vm-1.2-webapp/assets/21245760/17e55484-48ef-4716-9a66-6426f0d715d6)

4. the server disconnecting you should not kick you back to the vm list, instead it should attempt to reconnect at least a few times first. (just a suggestion)

elijahr2411 commented:
> 1. build times are horrendous, and it uses stupid amounts of memory/cpu/disk while building

I'm consistently unable to reproduce this. On my machine:
![image](https://github.com/computernewb/collab-vm-1.2-webapp/assets/62162399/90fc4914-d44e-4348-a67a-da0282866ed3)
On a fresh Windows 10 VM with nothing installed except latest NodeJS + build tools:
![image](https://github.com/computernewb/collab-vm-1.2-webapp/assets/62162399/e9bb5d0f-a116-46c2-9ead-392ae85e74c7)
So I'm pretty sure the problem here is on your end.

> 2. username does not persist (test: change username, then refresh)

fixed by [b60bdd9](https://github.com/computernewb/collab-vm-1.2-webapp/pull/16/commits/b60bdd9b65e66fe8fe96d75d10248309764724a1)

> 3. staff actions don't show up after logging in, and logging in makes username invisible

fixed by [a23f0e8](https://github.com/computernewb/collab-vm-1.2-webapp/pull/16/commits/a23f0e8886b2fd51869de9a2f167208a80b3e41d)

> 4. the server disconnecting you should not kick you back to the vm list, instead it should attempt to reconnect at least a few times first. (just a suggestion)

Having it auto-reconnect seems a bit redundant now that we actually have working VM switching. Additionally, kicking the user back to the VM list makes the Kick staff function not completely useless. This was my rationale for not implementing auto-reconnect but I'm happy to hear opinions on it
